README
= Ruby/SerialPort

== Description

Ruby/SerialPort is a Ruby library that provides a class for using
RS-232 serial ports.  This class also contains low-level functions to
check and set the current state of the signals on the line.

It support POSIX (Linux, *BSD, Solaris, AIX, Mac OS X), Cygwin and native Windows.
The native Windows version of this library supports Microsoft's Visual C++
and Borland's C++ compilers.

== Installation

$ gem install ruby-serialport

Any files and older versions can be found at http://rubyforge.org/projects/ruby-serialport/

== Examples

 reuire 'rubygems'
 require 'serialport'
 sp = SerialPort.new "/dev/my-modem", 9600
 sp.write "AT\r\n"
 puts sp.read   # hopefully "OK" ;-)

A functional example can be found in <tt>examples/miniterm.rb</tt> -- Ruby's copy of miniterm.c

For a compeate API reference see the RDoc of SerialPort.

== License and Copyright

 Copyright (C) 2008 Jonas Bähr <jonas.baehr@fs.ei.tum.de>
           (C) 2008 Daniel E. Shipton <daniel.shipton.oss@gmail.com> 
           (C) 2003 Alan Stern <stern@rowland.harvard.edu>
           (C) 2002 Guillaume Pierronnet <moumar@netcourrier.com>
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.



ChangeLog
=== 0.7.1 => ???
 * fix compile under ruby 1.9.1 (essentially applying patch from 
	http://makrotopia.org/browser/makrosys/makrobe/ruby-serialport/patches/10-fix-for-ruby-1.9)

=== 0.7.0 => 11/1/2009
 * Replaced handwritten Rake tasks and Gemspec with Echoe
 * Fully RDoc'ified
 * Support for Mac OS X
 * Major Code Cleanup
 * It's a proper RubyGem now

=== 0.6.1 => 25/03/2003
 * Minor changes

=== 0.6 => 12/02/2003
 * Windows support
 * Get/set modem parameters
 * Read and write timeouts
 * Open method

=== 0.5 => 25/10/2002
 * Cygwin support

=== 0.4 => 19/09/2002
 * Added more serial ports (total of 8 ports)

=== 0.3 => 15/03/2002
 * Damn, another bug found

=== 0.2 => 14/03/2002
 * A bug fixed (read() was not blocking)

=== 0.1 => 14/03/2002
 * First release


ext/serialport.h
/* Ruby/SerialPort $Id$
 * Guillaume Pierronnet <moumar@netcourrier.com>
 * Alan Stern <stern@rowland.harvard.edu>
 * Daniel E. Shipton <dshipton@redshiptechnologies.com>
 *
 * This code is hereby licensed for public consumption under either the
 * GNU GPL v2 or greater.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * For documentation on serial programming, see the excellent:
 * "Serial Programming Guide for POSIX Operating Systems"
 * written Michael R. Sweet.
 * http://www.easysw.com/~mike/serial/
 */

#ifndef _RUBY_SERIAL_PORT_H_
#define _RUBY_SERIAL_PORT_H_

#define RUBY_SERIAL_PORT_VERSION   "0.7.1"

#include <ruby.h>    /* ruby inclusion */
#include <ruby/io.h>  /* ruby io inclusion */

struct modem_params
{
   int data_rate;
   int data_bits;
   int stop_bits;
   int parity;
};

struct line_signals
{
   int rts;
   int dtr;
   int cts;
   int dsr;
   int dcd;
   int ri;
};

#define NONE   0
#define HARD   1
#define SOFT   2

#if defined(OS_MSWIN) || defined(OS_BCCWIN)
   #define SPACE  SPACEPARITY
   #define MARK   MARKPARITY
   #define EVEN   EVENPARITY
   #define ODD    ODDPARITY
#else
   #define SPACE  0
   #define MARK   0
   #define EVEN   1
   #define ODD    2
#endif

static VALUE sBaud, sDataBits, sStopBits, sParity; /* strings */
static VALUE sRts, sDtr, sCts, sDsr, sDcd, sRi;

#endif


lib/serialport.rb
require "serialport.so"

class SerialPort
   private_class_method(:create)

   # Creates a serial port object.
   #
   # <tt>port</tt> may be a port number
   # or the file name of a defice.
   # The number is portable; so 0 is mapped to "COM1" on Windows,
   # "/dev/ttyS0" on Linux, "/dev/cuaa0" on Mac OS X, etc.
   #
   # <tt>params</tt> can be used to configure the serial port.
   # See SerialPort#set_modem_params for details
   def SerialPort::new(port, *params)
      sp = create(port)
      begin
         sp.set_modem_params(*params)
      rescue
         sp.close
         raise
      end
      return sp
   end

   # This behaves like SerialPort#new, except that you can pass a block
   # to which the new serial port object will be passed. In this case
   # the connection is automaticaly closed when the block has finished.
   def SerialPort::open(port, *params)
      sp = create(port)
      begin
         sp.set_modem_params(*params)
         if (block_given?)
            yield sp
            sp.close
            return nil
         end
      rescue
         sp.close
         raise
      end
      return sp
   end
end


ext/serialport.c
/* Ruby/SerialPort $Id$
 * Guillaume Pierronnet <moumar@netcourrier.com>
 * Alan Stern <stern@rowland.harvard.edu>
 * Daniel E. Shipton <dshipton@redshiptechnologies.com>
 * Jonas Bähr <jonas.baehr@fs.ei.tum.de>
 *
 * This code is hereby licensed for public consumption under either the
 * GNU GPL v2 or greater.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * For documentation on serial programming, see the excellent:
 * "Serial Programming Guide for POSIX Operating Systems"
 * written Michael R. Sweet.
 * http://www.easysw.com/~mike/serial/
 */

#include "serialport.h"

/* include only the implementation we need */
#if defined(OS_MSWIN) || defined(OS_BCCWIN)
#include "impl/win_serialport.c"
#else
#include "impl/posix_serialport.c"
#endif /* defined(OS_MSWIN) || defined(OS_BCCWIN) */

VALUE cSerialPort; /* serial port class */

/*
 * :nodoc: This method is privat and will be called by SerialPort#new or SerialPort#open.
 */
static VALUE sp_create(class, _port)
   VALUE class, _port;
{
   return sp_create_impl(class, _port);
}

/*
 * Configure the serial port. You can pass a hash or multiple values
 * as separate arguments. Invalid or unsupported values will raise
 * an ArgumentError.
 *
 * When using a hash the following keys are recognized:
 * ["baud"] Integer from 50 to 256000, depends on platform
 * ["data_bits"] Integer from 5 to 8 (4 is allowed on Windows too)
 * ["stop_bits"] An integer, only allowed values are 1 or 2 (1.5 is not supported)
 * ["parity"] One of the constants NONE, EVEN or ODD (Windows allows also MARK and SPACE)
 *
 * When using separate arguments, they are interpreted as:
 * (baud, data_bits = 8, stop_bits = 1, parity = (previous_databits==8 ? NONE : EVEN))
 *
 * Nota: A baudrate of nil will keep the old value. The default parity depends on the
 * number of databits configured before this function call.
 */
static VALUE sp_set_modem_params(argc, argv, self)
   int argc;
   VALUE *argv, self;
{
   return sp_set_modem_params_impl(argc, argv, self);
}

/*
 * Send a break for the given time.
 *
 * <tt>time</tt> is an integer of tenths-of-a-second for the break.
 *
 * Note: Under Posix, this value is very approximate.
 */
static VALUE sp_break(self, time)
   VALUE self, time;
{
   return sp_break_impl(self, time);
}

/*
 * Get the state (0 or 1) of the DTR line (not available on Windows)
 */
static VALUE sp_get_dtr(self)
   VALUE self;
{
   return sp_get_dtr_impl(self);
}

/*
 * Get the flow control. The result is either NONE, HARD, SOFT or (HARD | SOFT)
 */
static VALUE sp_get_flow_control(self)
   VALUE self;
{
   return sp_get_flow_control_impl(self);
}

/*
 * Get the timeout value (in milliseconds) for reading.
 * See SerialPort#set_read_timeout for details.
 */
static VALUE sp_get_read_timeout(self)
   VALUE self;
{
   return sp_get_read_timeout_impl(self);
}

/*
 * Get the state (0 or 1) of the RTS line (not available on Windows)
 */
static VALUE sp_get_rts(self)
   VALUE self;
{
   return sp_get_rts_impl(self);
}

/*
 * Get the write timeout (in milliseconds)
 *
 * Note: Under Posix, write timeouts are not implemented.
 */
static VALUE sp_get_write_timeout(self)
   VALUE self;
{
   return sp_get_write_timeout_impl(self);
}

/*
 * Set the state (0 or 1) of the DTR line
 */
static VALUE sp_set_dtr(self, val)
{
   return sp_set_dtr_impl(self, val);
}

/*
 * Set the flow control to either NONE, HARD, SOFT or (HARD | SOFT)
 *
 *  Note: SerialPort::HARD mode is not supported on all platforms.
 *  SerialPort::HARD uses RTS/CTS handshaking; DSR/DTR is not
 *  supported.
 */
static VALUE sp_set_flow_control(self, val)
{
   return sp_set_flow_control_impl(self, val);
}

/*
 * Set the timeout value (in milliseconds) for reading.
 * A negative read timeout will return all the available data without
 * waiting, a zero read timeout will not return until at least one
 * byte is available, and a positive read timeout returns when the
 * requested number of bytes is available or the interval between the
 * arrival of two bytes exceeds the timeout value.
 *
 * Note: Read timeouts don't mix well with multi-threading.
 */
static VALUE sp_set_read_timeout(self, val)
{
   return sp_set_read_timeout_impl(self, val);
}

/*
 * Set the state (0 or 1) of the RTS line
 */
static VALUE sp_set_rts(self, val)
{
   return sp_set_rts_impl(self, val);
}

/*
 * Set a write timeout (in milliseconds)
 *
 * Note: Under Posix, write timeouts are not implemented.
 */
static VALUE sp_set_write_timeout(self, val)
{
   return sp_set_write_timeout_impl(self, val);
}

static void get_modem_params(self, mp)
   VALUE self;
   struct modem_params *mp;
{
   get_modem_params_impl(self, mp);
}

/*
 * Set the baud rate, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_data_rate(self, data_rate)
   VALUE self, data_rate;
{
   VALUE argv[4];

   argv[0] = data_rate;
   argv[1] = argv[2] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return data_rate;
}

/*
 * Set the data bits, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_data_bits(self, data_bits)
   VALUE self, data_bits;
{
   VALUE argv[4];

   argv[1] = data_bits;
   argv[0] = argv[2] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return data_bits;
}

/*
 * Set the stop bits, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_stop_bits(self, stop_bits)
   VALUE self, stop_bits;
{
   VALUE argv[4];

   argv[2] = stop_bits;
   argv[0] = argv[1] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return stop_bits;
}

/*
 * Set the parity, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_parity(self, parity)
   VALUE self, parity;
{
   VALUE argv[4];

   argv[3] = parity;
   argv[0] = argv[1] = argv[2] = Qnil;
   sp_set_modem_params(4, argv, self);

   return parity;
}

/*
 * Get the current baud rate, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_data_rate(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.data_rate);
}

/*
 * Get the current data bits, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_data_bits(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.data_bits);
}

/*
 * Get the current stop bits, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_stop_bits(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.stop_bits);
}

/*
 * Get the current parity, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_parity(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.parity);
}


/*
 * Get the configure of the serial port.
 *
 * Returned is a hash with the following keys:
 * ["baud"] Integer with the baud rate
 * ["data_bits"] Integer from 5 to 8 (4 is possible on Windows too)
 * ["stop_bits"] Integer, 1 or 2 (1.5 is not supported)
 * ["parity"] One of the constants NONE, EVEN or ODD (on Windows may also MARK or SPACE)
 */
static VALUE sp_get_modem_params(self)
   VALUE self;
{
   struct modem_params mp;
   VALUE hash;

   get_modem_params(self, &mp);

   hash = rb_hash_new();

   rb_hash_aset(hash, sBaud, INT2FIX(mp.data_rate));
   rb_hash_aset(hash, sDataBits, INT2FIX(mp.data_bits));
   rb_hash_aset(hash, sStopBits, INT2FIX(mp.stop_bits));
   rb_hash_aset(hash, sParity, INT2FIX(mp.parity));

   return hash;
}

/*
 * Get the state (0 or 1) of the CTS line
 */
static VALUE sp_get_cts(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.cts);
}

/*
 * Get the state (0 or 1) of the DSR line
 */
static VALUE sp_get_dsr(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.dsr);
}

/*
 * Get the state (0 or 1) of the DCD line
 */
static VALUE sp_get_dcd(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.dcd);
}

/*
 * Get the state (0 or 1) of the RI line
 */
static VALUE sp_get_ri(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.ri);
}

/*
 * Return a hash with the state of each line status bit.  Keys are
 * "rts", "dtr", "cts", "dsr", "dcd", and "ri".
 *
 * Note: Under Windows, the rts and dtr values are not included.
 */
static VALUE sp_signals(self)
   VALUE self;
{
   struct line_signals ls;
   VALUE hash;

   get_line_signals_helper(self, &ls);

   hash = rb_hash_new();

#if !(defined(OS_MSWIN) || defined(OS_BCCWIN))
   rb_hash_aset(hash, sRts, INT2FIX(ls.rts));
   rb_hash_aset(hash, sDtr, INT2FIX(ls.dtr));
#endif
   rb_hash_aset(hash, sCts, INT2FIX(ls.cts));
   rb_hash_aset(hash, sDsr, INT2FIX(ls.dsr));
   rb_hash_aset(hash, sDcd, INT2FIX(ls.dcd));
   rb_hash_aset(hash, sRi, INT2FIX(ls.ri));

   return hash;
}

/*
 * This class is used for communication over a serial port.
 * In addition to the methods here, you can use everything
 * Ruby's IO-class provides (read, write, getc, readlines, ...)
 */
void Init_serialport()
{
   sBaud = rb_str_new2("baud");
   sDataBits = rb_str_new2("data_bits");
   sStopBits = rb_str_new2("stop_bits");
   sParity = rb_str_new2("parity");
   sRts = rb_str_new2("rts");
   sDtr = rb_str_new2("dtr");
   sCts = rb_str_new2("cts");
   sDsr = rb_str_new2("dsr");
   sDcd = rb_str_new2("dcd");
   sRi = rb_str_new2("ri");

   rb_gc_register_address(&sBaud);
   rb_gc_register_address(&sDataBits);
   rb_gc_register_address(&sStopBits);
   rb_gc_register_address(&sParity);
   rb_gc_register_address(&sRts);
   rb_gc_register_address(&sDtr);
   rb_gc_register_address(&sCts);
   rb_gc_register_address(&sDsr);
   rb_gc_register_address(&sDcd);
   rb_gc_register_address(&sRi);

   cSerialPort = rb_define_class("SerialPort", rb_cIO);
   rb_define_singleton_method(cSerialPort, "create", sp_create, 1);

   rb_define_method(cSerialPort, "get_modem_params", sp_get_modem_params, 0);
   rb_define_method(cSerialPort, "set_modem_params", sp_set_modem_params, -1);
   rb_define_method(cSerialPort, "modem_params", sp_get_modem_params, 0);
   rb_define_method(cSerialPort, "modem_params=", sp_set_modem_params, -1);
   rb_define_method(cSerialPort, "baud", sp_get_data_rate, 0);
   rb_define_method(cSerialPort, "baud=", sp_set_data_rate, 1);
   rb_define_method(cSerialPort, "data_bits", sp_get_data_bits, 0);
   rb_define_method(cSerialPort, "data_bits=", sp_set_data_bits, 1);
   rb_define_method(cSerialPort, "stop_bits", sp_get_stop_bits, 0);
   rb_define_method(cSerialPort, "stop_bits=", sp_set_stop_bits, 1);
   rb_define_method(cSerialPort, "parity", sp_get_parity, 0);
   rb_define_method(cSerialPort, "parity=", sp_set_parity, 1);

   rb_define_method(cSerialPort, "flow_control=", sp_set_flow_control, 1);
   rb_define_method(cSerialPort, "flow_control", sp_get_flow_control, 0);

   rb_define_method(cSerialPort, "read_timeout", sp_get_read_timeout, 0);
   rb_define_method(cSerialPort, "read_timeout=", sp_set_read_timeout, 1);
   rb_define_method(cSerialPort, "write_timeout", sp_get_write_timeout, 0);
   rb_define_method(cSerialPort, "write_timeout=", sp_set_write_timeout, 1);

   rb_define_method(cSerialPort, "break", sp_break, 1);

   rb_define_method(cSerialPort, "signals", sp_signals, 0);
   rb_define_method(cSerialPort, "get_signals", sp_signals, 0);
   rb_define_method(cSerialPort, "rts", sp_get_rts, 0);
   rb_define_method(cSerialPort, "rts=", sp_set_rts, 1);
   rb_define_method(cSerialPort, "dtr", sp_get_dtr, 0);
   rb_define_method(cSerialPort, "dtr=", sp_set_dtr, 1);
   rb_define_method(cSerialPort, "cts", sp_get_cts, 0);
   rb_define_method(cSerialPort, "dsr", sp_get_dsr, 0);
   rb_define_method(cSerialPort, "dcd", sp_get_dcd, 0);
   rb_define_method(cSerialPort, "ri", sp_get_ri, 0);

   rb_define_const(cSerialPort, "NONE", INT2FIX(NONE));
   rb_define_const(cSerialPort, "HARD", INT2FIX(HARD));
   rb_define_const(cSerialPort, "SOFT", INT2FIX(SOFT));

   rb_define_const(cSerialPort, "SPACE", INT2FIX(SPACE));
   rb_define_const(cSerialPort, "MARK", INT2FIX(MARK));
   rb_define_const(cSerialPort, "EVEN", INT2FIX(EVEN));
   rb_define_const(cSerialPort, "ODD", INT2FIX(ODD));

   /* the package's version as a string "X.Y.Z", beeing major, minor and patch level */
   rb_define_const(cSerialPort, "VERSION", rb_str_new2(RUBY_SERIAL_PORT_VERSION));
}
